drivers/ina3221/ina3221.c

```diff
@@ -0,0 +1,928 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ina3221
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for Texas Instruments INA3221
+ *              three-channel, high-side current and bus voltage
+ *              monitor
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <periph/gpio.h>
+#include <byteorder.h>
+#define ENABLE_DEBUG (1)
+#include "debug.h"
+
+#include "include/ina3221_defines.h"
+#include "include/ina3221_params.h"
+#include "include/ina3221_regs.h"
+#include "ina3221.h"
+
+struct channel_info
+{
+    uint8_t  chi_reg_shunt;
+    uint8_t  chi_reg_bus;
+    uint8_t  chi_reg_crit_alert_limit;
+    uint8_t  chi_reg_warn_alert_limit;
+};
+
+static struct channel_info _chi[INA3221_NUM_CH] = {
+    (struct channel_info){
+        .chi_reg_shunt            = INA3221_REG_CH1_SHUNT_VOLTAGE,
+        .chi_reg_bus              = INA3221_REG_CH1_BUS_VOLTAGE,
+        .chi_reg_crit_alert_limit = INA3221_REG_CH1_CRIT_ALERT_LIMIT,
+        .chi_reg_warn_alert_limit = INA3221_REG_CH1_WARN_ALERT_LIMIT
+    },
+    (struct channel_info){
+        .chi_reg_shunt            = INA3221_REG_CH2_SHUNT_VOLTAGE,
+        .chi_reg_bus              = INA3221_REG_CH2_BUS_VOLTAGE,
+        .chi_reg_crit_alert_limit = INA3221_REG_CH2_CRIT_ALERT_LIMIT,
+        .chi_reg_warn_alert_limit = INA3221_REG_CH2_WARN_ALERT_LIMIT
+    },
+    (struct channel_info){
+        .chi_reg_shunt            = INA3221_REG_CH3_SHUNT_VOLTAGE,
+        .chi_reg_bus              = INA3221_REG_CH3_BUS_VOLTAGE,
+        .chi_reg_crit_alert_limit = INA3221_REG_CH3_CRIT_ALERT_LIMIT,
+        .chi_reg_warn_alert_limit = INA3221_REG_CH3_WARN_ALERT_LIMIT
+    },
+};
+
+/**
+ * @brief Convert register value to shunt voltage in uV
+ * 
+ * @param[in]   reg_val Register value
+ * 
+ * @pre         reg_val must be in host byte order
+ * 
+ * @returm      Shunt voltage in uV
+ */
+static inline int32_t reg_val_to_shunt_voltage_uV(int16_t reg_val)
+{
+    //assert(reg_val >= INA3221_MIN_SHUNT_REG_VAL);
+    assert(reg_val <= INA3221_MAX_SHUNT_REG_VAL);
+    
+    int32_t shunt_voltage_uV;
+    if(reg_val < 0)
+    {
+        shunt_voltage_uV = (~(reg_val) + 1) / 8 * INA3221_SHUNT_VOLTAGE_PRECISION_uV;
+        shunt_voltage_uV *= -1;
+    }
+    else
+    {
+        shunt_voltage_uV =  reg_val / 8 * INA3221_SHUNT_VOLTAGE_PRECISION_uV;
+    }
```

Comment:
Trying to avoid signed division, as unsigned division is more efficient on some platforms?

This will likely not help, as the data type still is signed. It is indeed better to avoid signed division by choosing an `unsigned` data type instead of a `signed` if this is possible. Otherwise, just go for signed division and readable code and let the compiler to the optimization.

(Also not that changing the sign of an `x` can be done most readable and efficient with the unary `-` operator; e.g. `x = -x;` would change it. Depending on ISA there also is a single instruction to change the sign.)

Reply:
Oh interesting!
Actually I just thought a negative `reg_val` needs special treatment, but it appears to be the same.^^

However, do you think this avoids signed division? (I will test it later.)
Or do you think avoiding signed division is not worth it?
```
    uint32_t shunt_voltage_uV;
    if(reg_val < 0)
    {
        shunt_voltage_uV = (~((uint16_t)reg_val) + 1U) / 8U * INA3221_SHUNT_VOLTAGE_PRECISION_uV;
        shunt_voltage_uV = (~shunt_voltage_uV) + 1U;  
    }
    else
    {
        shunt_voltage_uV =  ((uint16_t)reg_val) / 8U * INA3221_SHUNT_VOLTAGE_PRECISION_uV;
    }
    return (int32_t)shunt_voltage_uV;
```

Reply:
My experience with trying to optimize that low level in C is that the compiler does a better job with low level optimization anyway. (Or maybe I just suck at low level optimization :-D) But you can check `make info-buildsize` or disassemble the function. [On armv6 unsigned division should be smaller](https://github.com/RIOT-OS/RIOT/pull/9702#pullrequestreview-160609172)

Anyways, the code will be much easier to read when you just use signed division. To me personally that would be more important.

If this indeed safes >= 100B of ROM, you could create a helper function for this such as

``` C
static inline int16_t div_s16(int16_t dividend, uint16_t divisor)
{
    if (dividiend >= 0) {
        return (int16_t)(((uint16_t)dividend) / divisor);
    }
    return (int16_t)((uint16_t)(-dividend) / divisor);
}
```

to improve readability a bit and still have the smaller ROM size.

Reply:
normal:
18348     120    2728   21196    52cc /home/fabian/forks/RIOT/tests/driver_ina3221/bin/nucleo-f446re/tests_driver_ina3221.elf

avoid signed division:
18444     120    2728   21292    532c /home/fabian/forks/RIOT/tests/driver_ina3221/bin/nucleo-f446re/tests_driver_ina3221.elf

So the unsigned version takes even more ROM.